File: objectivec/day4_part1_2022.m
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *input = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:nil];
        NSArray *lines = [input componentsSeparatedByString:@"\n"];
        
        int count = 0;
        
        for (NSString *line in lines) {
            NSArray *sections = [line componentsSeparatedByString:@","];
            NSArray *range1 = [sections[0] componentsSeparatedByString:@"-"];
            NSArray *range2 = [sections[1] componentsSeparatedByString:@"-"];
            
            int start1 = [range1[0] intValue];
            int end1 = [range1[1] intValue];
            int start2 = [range2[0] intValue];
            int end2 = [range2[1] intValue];
            
            if ((start1 <= start2 && end1 >= end2) || (start2 <= start1 && end2 >= end1)) {
                count++;
            }
        }
        
        printf("%d\n", count);
    }
    return 0;
}

File: fortran90/day4_part1_2022.f90

program main
  implicit none
  integer :: count, start1, end1, start2, end2
  character(len=20) :: line
  integer, dimension(2) :: range1, range2
  integer :: iostat
  open(unit=10, file="input.txt", status="old", action="read", iostat=iostat)
  if (iostat /= 0) then
    print *, "Error opening file"
    stop
  end if
  count = 0
  do
    read(10, '(a)', iostat=iostat) line
    if (iostat /= 0) exit
    call parse_range(line, range1, range2)
    start1 = range1(1)
    end1 = range1(2)
    start2 = range2(1)
    end2 = range2(2)
    if ((start1 <= start2 .and. end1 >= end2) .or. (start2 <= start1 .and. end2 >= end1)) then
      count = count + 1
    end if
  end do
  close(10)
  print *, count
contains
  subroutine parse_range(line, range1, range2)
    implicit none
    character(len=*), intent(in) :: line
    integer, dimension(2), intent(out) :: range1, range2
    integer :: i, j, k, dash_pos, comma_pos
    character(len=20) :: temp
    comma_pos = index(line, ",")
    temp = line(1:comma_pos-1)
    dash_pos = index(temp, "-")
    read(temp(1:dash_pos-1), *) range1(1)
    read(temp(dash_pos+1:), *) range1(2)
    temp = line(comma_pos+1:)
    dash_pos = index(temp, "-")
    read(temp(1:dash_pos-1), *) range2(1)
    read(temp(dash_pos+1:), *) range2(2)
  end subroutine parse_range
end program main
